// braintree/lib/braintree/braintree_gateway.js
// Generated by CoffeeScript 1.6.1
var AddOnGateway, AddressGateway, BraintreeGateway, ClientTokenGateway, CreditCardGateway, CreditCardVerificationGateway, CustomerGateway, DisbursementGateway, DiscountGateway, Http, MerchantAccountGateway, MerchantGateway, OAuthGateway, PayPalAccountGateway, PaymentMethodGateway, PaymentMethodNonceGateway, PlanGateway, SettlementBatchSummaryGateway, SubscriptionGateway, TestingGateway, TransactionGateway, TransparentRedirectGateway, UsBankAccountGateway, WebhookNotificationGateway, WebhookTestingGateway;

Http = require('./http').Http;

AddOnGateway = require("./add_on_gateway").AddOnGateway;

AddressGateway = require("./address_gateway").AddressGateway;

ClientTokenGateway = require("./client_token_gateway").ClientTokenGateway;

CreditCardGateway = require("./credit_card_gateway").CreditCardGateway;

CreditCardVerificationGateway = require("./credit_card_verification_gateway").CreditCardVerificationGateway;

CustomerGateway = require("./customer_gateway").CustomerGateway;

DisbursementGateway = require("./disbursement_gateway").DisbursementGateway;

DiscountGateway = require("./discount_gateway").DiscountGateway;

MerchantAccountGateway = require("./merchant_account_gateway").MerchantAccountGateway;

MerchantGateway = require("./merchant_gateway").MerchantGateway;

OAuthGateway = require("./oauth_gateway").OAuthGateway;

PaymentMethodGateway = require("./payment_method_gateway").PaymentMethodGateway;

PaymentMethodNonceGateway = require("./payment_method_nonce_gateway").PaymentMethodNonceGateway;

PayPalAccountGateway = require("./paypal_account_gateway").PayPalAccountGateway;

PlanGateway = require("./plan_gateway").PlanGateway;

SettlementBatchSummaryGateway = require("./settlement_batch_summary_gateway").SettlementBatchSummaryGateway;

SubscriptionGateway = require("./subscription_gateway").SubscriptionGateway;

TestingGateway = require("./testing_gateway").TestingGateway;

TransactionGateway = require("./transaction_gateway").TransactionGateway;

TransparentRedirectGateway = require("./transparent_redirect_gateway").TransparentRedirectGateway;

UsBankAccountGateway = require("./us_bank_account_gateway").UsBankAccountGateway;

WebhookNotificationGateway = require("./webhook_notification_gateway").WebhookNotificationGateway;

WebhookTestingGateway = require("./webhook_testing_gateway").WebhookTestingGateway;

BraintreeGateway = (function() {

  function BraintreeGateway(config) {
    this.config = config;
    this.http = new Http(this.config);
    this.addOn = new AddOnGateway(this);
    this.address = new AddressGateway(this);
    this.clientToken = new ClientTokenGateway(this);
    this.creditCard = new CreditCardGateway(this);
    this.creditCardVerification = new CreditCardVerificationGateway(this);
    this.customer = new CustomerGateway(this);
    this.disbursement = new DisbursementGateway(this);
    this.discount = new DiscountGateway(this);
    this.merchantAccount = new MerchantAccountGateway(this);
    this.merchant = new MerchantGateway(this);
    this.oauth = new OAuthGateway(this);
    this.paymentMethod = new PaymentMethodGateway(this);
    this.paymentMethodNonce = new PaymentMethodNonceGateway(this);
    this.paypalAccount = new PayPalAccountGateway(this);
    this.plan = new PlanGateway(this);
    this.settlementBatchSummary = new SettlementBatchSummaryGateway(this);
    this.subscription = new SubscriptionGateway(this);
    this.testing = new TestingGateway(this);
    this.transaction = new TransactionGateway(this);
    this.transparentRedirect = new TransparentRedirectGateway(this);
    this.usBankAccount = new UsBankAccountGateway(this);
    this.webhookNotification = new WebhookNotificationGateway(this);
    this.webhookTesting = new WebhookTestingGateway(this);
  }

  return BraintreeGateway;

})();

exports.BraintreeGateway = BraintreeGateway;
