// braintree/lib/braintree/coinbase_account_gateway.js
// Generated by CoffeeScript 1.6.1
var CoinbaseAccount, CoinbaseAccountGateway, Gateway, exceptions,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

Gateway = require('./gateway').Gateway;

CoinbaseAccount = require('./coinbase_account').CoinbaseAccount;

exceptions = require('./exceptions');

CoinbaseAccountGateway = (function(_super) {

  __extends(CoinbaseAccountGateway, _super);

  function CoinbaseAccountGateway(gateway) {
    this.gateway = gateway;
    this.config = this.gateway.config;
  }

  CoinbaseAccountGateway.prototype.find = function(token, callback) {
    if (token.trim() === '') {
      return callback(exceptions.NotFoundError("Not Found"), null);
    } else {
      return this.gateway.http.get("" + (this.config.baseMerchantPath()) + "/payment_methods/coinbase_account/" + token, function(err, response) {
        if (err) {
          return callback(err, null);
        } else {
          return callback(null, new CoinbaseAccount(response.coinbaseAccount));
        }
      });
    }
  };

  CoinbaseAccountGateway.prototype["delete"] = function(token, callback) {
    return this.gateway.http["delete"]("" + (this.config.baseMerchantPath()) + "/payment_methods/coinbase_account/" + token, callback);
  };

  CoinbaseAccountGateway.prototype.responseHandler = function(callback) {
    return this.createResponseHandler("coinbaseAccount", CoinbaseAccount, callback);
  };

  return CoinbaseAccountGateway;

})(Gateway);

exports.CoinbaseAccountGateway = CoinbaseAccountGateway;
