// braintree/lib/braintree/credit_card_verification.js
// Generated by CoffeeScript 1.6.1
var AttributeSetter, CreditCardVerification, RiskData,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

AttributeSetter = require('./attribute_setter').AttributeSetter;

RiskData = require('./risk_data').RiskData;

CreditCardVerification = (function(_super) {

  __extends(CreditCardVerification, _super);

  CreditCardVerification.StatusType = {
    Failed: "failed",
    GatewayRejected: "gateway_rejected",
    ProcessorDeclined: "processor_declined",
    Verified: "verified",
    All: function() {
      var all, key, value;
      all = [];
      for (key in this) {
        value = this[key];
        if (key !== 'All') {
          all.push(value);
        }
      }
      return all;
    }
  };

  function CreditCardVerification(attributes) {
    CreditCardVerification.__super__.constructor.call(this, attributes);
    if (attributes.riskData) {
      this.riskData = new RiskData(attributes.riskData);
    }
  }

  return CreditCardVerification;

})(AttributeSetter);

exports.CreditCardVerification = CreditCardVerification;
