// braintree/lib/braintree/dispute.js
// Generated by CoffeeScript 1.6.1
var AttributeSetter, Dispute, TransactionDetails,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

AttributeSetter = require('./attribute_setter').AttributeSetter;

TransactionDetails = require('./transaction_details').TransactionDetails;

Dispute = (function(_super) {

  __extends(Dispute, _super);

  Dispute.Status = {
    Open: 'open',
    Lost: 'lost',
    Won: 'won'
  };

  Dispute.Reason = {
    CancelledRecurringTransaction: 'cancelled_recurring_transaction',
    CreditNotProcessed: 'credit_not_processed',
    Duplicate: 'duplicate',
    Fraud: 'fraud',
    General: 'general',
    InvalidAccount: 'invalid_account',
    NotRecognized: 'not_recognized',
    ProductNotReceived: 'product_not_received',
    ProductUnsatisfactory: 'product_unsatisfactory',
    Retrieval: 'retrieval',
    TransactionAmountDiffers: 'transaction_amount_differs'
  };

  Dispute.Kind = {
    Chargeback: 'chargeback',
    PreArbitration: 'pre_arbitration',
    Retrieval: 'retrieval'
  };

  function Dispute(attributes) {
    Dispute.__super__.constructor.call(this, attributes);
    this.transactionDetails = new TransactionDetails(attributes.transaction);
  }

  return Dispute;

})(AttributeSetter);

exports.Dispute = Dispute;
