// braintree/lib/braintree/merchant_account.js
// Generated by CoffeeScript 1.6.1
var AttributeSetter, MerchantAccount,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

AttributeSetter = require('./attribute_setter').AttributeSetter;

MerchantAccount = (function(_super) {

  __extends(MerchantAccount, _super);

  MerchantAccount.Status = {
    Pending: "pending",
    Active: "active",
    Suspended: "suspended"
  };

  MerchantAccount.FundingDestination = {
    Bank: "bank",
    Email: "email",
    MobilePhone: "mobile_phone"
  };

  function MerchantAccount(attributes) {
    MerchantAccount.__super__.constructor.call(this, attributes);
    if (attributes.masterMerchantAccount) {
      this.masterMerchantAccount = new MerchantAccount(attributes.masterMerchantAccount);
    }
  }

  return MerchantAccount;

})(AttributeSetter);

exports.MerchantAccount = MerchantAccount;
